### PR TITLE
Refactor award_yearly_club_badges

### PR DESCRIPTION
**Pull request by Glennmen**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [ ] Documentation Update

## Description
Refactor `award_yearly_club_badges` so that adding new yearly badges will be easier and require less changes.

I got a couple of questions about the refactor:

- Do you want it in a loop or more like `award_streak_badge(num_weeks)`? I think the for loop is a better option, you will only need to increase a variable and not add a new line to `fetch.rake`
https://github.com/thepracticaldev/dev.to/blob/0b84ae9acba3658af1112aeb520a049e852ffd27/app/labor/badge_rewarder.rb#L47
https://github.com/thepracticaldev/dev.to/blob/0b84ae9acba3658af1112aeb520a049e852ffd27/lib/tasks/fetch.rake#L73-L75

- What do you guys think is the best way to set the parameter? I currently added it as a variable in the function.

- Will changing the slug to a numeric number value instead of the word value (`one-year-club` -> `1-year-club` break existing badges? How to best handle this in your opinion?

Any other feedback is welcome!

## Related Tickets & Documents
https://dev.to/ben/comment/8inc

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

birthday

### CODE REVIEW

File: app/labor/badge_rewarder.rb [hunk 1]
```diff
@@ -1,8 +1,17 @@
 module BadgeRewarder
   def self.award_yearly_club_badges
-    award_one_year_badges
-    award_two_year_badges
-    award_three_year_badges
+    years = 3
+    (1..years).each do |i|
+      message = "Happy DEV birthday! Can you believe it's been #{i} years already?!"
```
Comment (abraham) [inline app/labor/badge_rewarder.rb]: The pluralization of `year` will be wrong for the first year.

> Happy DEV birthday! Can you believe it's been 1 years already?!

```suggestion
      message = "Happy DEV birthday! Can you believe it's been #{i} #{"year".pluralize(i)} already?!"
```

Comment (Glennmen) [reply app/labor/badge_rewarder.rb]: @abraham I didn't even think about this, thank you!

File: app/labor/badge_rewarder.rb [hunk 2]
```diff
@@ -1,8 +1,17 @@
...
+      User.where("created_at < ? AND created_at > ?", i.year.ago, i.year.ago - 2.day).find_each do |user|
+        achievement = BadgeAchievement.create(
+          user_id: user.id,
+          badge_id: Badge.find_by_slug("#{i}-year-club").id,
```
Comment (abraham) [inline app/labor/badge_rewarder.rb]: This looks like it'll run the `Badge.find_by_slug("#{i}-year-club")` query for each user found. I would probably move that query next to where `message` is defined to have it run once.

```ruby
    (1..years).each do |i|
      message = "Happy DEV birthday! Can you believe it's been #{i} years already?!"
      badge = Badge.find_by_slug("#{i}-year-club")
```

Then you reference the same badge instance for every iteration of that year.

```ruby
        achievement = BadgeAchievement.create(
          user_id: user.id,
          badge_id: badge.id,
```

Comment (Glennmen) [reply app/labor/badge_rewarder.rb]: I used the already available code but indeed moving it out of the query loop would be better.

Comment (abraham) [reply app/labor/badge_rewarder.rb]: Actually you should be able to simplify it even more and rely on Rails association handling.

```ruby
        achievement = BadgeAchievement.create(
          user: user,
          badge: badge,
```

Comment (Glennmen) [reply app/labor/badge_rewarder.rb]: @abraham If I remove `.id` the tests fail.

Comment (abraham) [reply app/labor/badge_rewarder.rb]: Hmmm...I wouldn't expect that to cause a test failure. Feel free to cleanup that minor cleanup suggestion or I can take a look later today.

Comment (Glennmen) [reply app/labor/badge_rewarder.rb]: I will add it now with `.id`

**Comment by lightalloy** [comment]
Great refactoring, @Glennmen 
I would also like smth like `years = Time.current.year - 2016` (not sure about the exact value), not to alter the number each year.

**Comment by Glennmen** [comment]
> Great refactoring, @Glennmen
> I would also like smth like `years = Time.current.year - 2016` (not sure about the exact value), not to alter the number each year.

@lightalloy I thought it would be better to have control over the value, for example to make sure that the icons are made for each year.

**Comment by lightalloy** [comment]
Returned to make a notice about icons ) (and badges records)
Now I think the dynamic badges shouldn't be a part of this pull request.

**Comment by Glennmen** [comment]
> Returned to make a notice about icons ) (and badges records)
> Now I think the dynamic badges shouldn't be a part of this pull request.

@lightalloy Sorry but I am not sure what you mean with this.

I was talking about having control over `years` to make sure that the badge icons will be available. Not that it will auto increase each year and DEV will forget to make the badge icons on time.

**Comment by lightalloy** [comment]
@Glennmen,  I realized there's a need to create a separate icon for each year and the `Badge` record for it in the database too (but only after commenting). Regarding this, I guess that this process should remain manual at the moment. So I suggest keeping your pull request as it is.

**Comment by Glennmen** [comment]
My 3 questions in the PR Description still remain open, just want to check everything before removing Draft status.

**Comment by lightalloy** [comment]
I looked a bit deeper into this and realized that changing slugs will require updating `Badge` records in the database.
I suggest keeping the old badge records and slugs. The possible solution is creating a constant like this (approximate code):
```ruby
YEARS = { 1 => "one", 2 => "two", 3 => "three"}.freeze
# ...
# in your method
YEARS.each do |i, word|
  # your code
end
```

**Comment by Glennmen** [comment]
Sorry for the delay, we had a long team building weekend in Austria with work.

I will change it to use the old slugs. It would be better if we could use decimal values to make it better long term but that would be for a different PR (maybe with something like this: https://github.com/radar/humanize).

Still the 2 other questions remain open, how does the DEV team want the function to look like.

- Do you want it in a loop or more like `award_streak_badge(num_weeks)`?
- What do you guys think is the best way to set the parameter? A fixed parameter in code or using env variable?

**Comment by lightalloy** [comment]
@Glennmen Thanks for returning to this issue :sun_with_face: 
- I would leave the decision to you
- Could you specify what parameter are you talking about?
If it's about years, I still suggest my approach from the above comment (a constant).

**Comment by Glennmen** [comment]
@lightalloy Yes the `years` constant.

Maybe @benhalpern can give his opinion on this one too. He suggested the environment variable.
https://dev.to/ben/comment/8inc

File: app/labor/badge_rewarder.rb [hunk 3]
```diff
@@ -1,8 +1,19 @@
 module BadgeRewarder
+  YEARS = { 1 => "one", 2 => "two", 3 => "three", 4 => "four", 5 => "five", 6 => "six", 7 => "seven", 8 => "eight", 9 => "nine", 10 => "ten" }.freeze
+
   def self.award_yearly_club_badges
-    award_one_year_badges
-    award_two_year_badges
-    award_three_year_badges
+    (1..3).each do |i|
```
Comment (Glennmen) [inline app/labor/badge_rewarder.rb]: Not sure how the DEV team would like this, to keep it easy for future updates. I don't like the idea of having it "auto update" based on current year.

Comment (maestromac) [reply app/labor/badge_rewarder.rb]: I agree. Let's not have it auto-update yet.

Comment (Glennmen) [reply app/labor/badge_rewarder.rb]: @maestromac So the current state is good to remove WIP? Or should `(1..3).each` be changed to an environment variable?
Will need to figure out how to do that, but I think I have a basic idea how.

Comment (maestromac) [reply app/labor/badge_rewarder.rb]: Yep you can remove the WIP. It's fine to keep it as is.

**Comment by Glennmen** [comment]
Strange, Travis doesn't get triggered if Draft has been removed.

**Comment by maestromac** [comment]
Going to trigger the build by closing and reopening

**Pull request merged**